verification/tpm.go

```diff
@@ -0,0 +1,268 @@
+		"loaded":    {tpm2.HandleTypeLoadedSession},
+		"saved":     {tpm2.HandleTypeSavedSession},
+		"transient": {tpm2.HandleTypeTransient},
+	}
+)
+
+func TestTpmPolicySigning(d TestDPEInstance, c DPEClient, t *testing.T) {
+	if strings.ToLower(*tpmPolicySigningValidation) == "enabled" {
+		t.Log("Validation of TPM policy signing using Caliptra DPE commands is enabled.")
+		testTpmPolicySigning(d, c, t)
+	} else {
+		t.Log("Validation of TPM policy signing using Caliptra DPE commands is disabled, skipping this test.\n To enable this validation start the TPM emulator, set `go test . -tpm-policy-signing-validation=enabled` and rerun.")
+		return
+	}
+}
+
+func startTpmSession(t *testing.T, tpm io.ReadWriteCloser, alg tpm2.Algorithm) (tpmutil.Handle, []byte, error) {
+	totalHandles := 0
+	for _, handleType := range handleNames["all"] {
+		handles, err := client.Handles(tpm, handleType)
+		if err != nil {
+			t.Fatalf("[FATAL]: Error getting handles %s: %v", *tpmPath, err)
+		}
+		for _, handle := range handles {
+			if err = tpm2.FlushContext(tpm, handle); err != nil {
+				t.Fatalf("[FATAL]: Error flushing handle 0x%x: %v", handle, err)
+			}
+			t.Logf("Handle 0x%x flushed", handle)
+			totalHandles++
+		}
+	}
+
+	sessHandle, nonce, err := tpm2.StartAuthSession(tpm,
+		tpm2.HandleNull,
+		tpm2.HandleNull,
+		make([]byte, 16),
+		nil,
+		tpm2.SessionPolicy,
+		tpm2.AlgNull,
+		alg)
+
+	if err != nil {
+		t.Fatalf("[FATAL]: StartAuthSession() failed: %v", err)
+	}
+
+	return sessHandle, nonce, nil
+}
+
+func testTpmPolicySigning(d TestDPEInstance, c DPEClient, t *testing.T) {
+	var ctx ContextHandle = [16]byte{0}
+	var ec tpm2.EllipticCurve
+	var alg tpm2.Algorithm
+
+	profile, err := GetTransportProfile(d)
+	if err != nil {
+		t.Fatalf("Could not get profile: %v", err)
+	}
+	digestLen := profile.GetDigestSize()
+
+	if digestLen == len(SHA256Digest{0}) {
+		alg = tpm2.AlgSHA256
+		ec = tpm2.CurveNISTP256
+	} else if digestLen == len(SHA384Digest{0}) {
+		alg = tpm2.AlgSHA384
+		ec = tpm2.CurveNISTP384
+	}
+
+	//Create tpm auth session to get nonce and form label which is digest
+	tpm, err := tpm2.OpenTPM(*tpmPath)
+	if err != nil {
+		t.Fatalf("[FATAL]: Can't open TPM %s: %v", *tpmPath, err)
+	}
+
+	sessHandle, nonce, err := startTpmSession(t, tpm, alg)
+	if err != nil {
+		t.Fatalf("[FATAL]: Error in getting tpm nonce")
+	}
+
+	defer func() {
+		if err := tpm.Close(); err != nil {
+			t.Fatalf("[FATAL]: Can't close TPM %s: %v", *tpmPath, err)
+		}
+	}()
+
+	defer tpm2.FlushContext(tpm, sessHandle)
+
+	// Build SignHash request
+	expiry := int32(math.MinInt32)
+	digest := getDigest(nonce, expiry, digestLen)
+
+	seqLabel := make([]byte, digestLen)
+	for i, _ := range seqLabel {
+		seqLabel[i] = byte(i)
+	}
+
+	signParams := struct {
+		Flags      SignFlags
+		Label      []byte
+		ToBeSigned []byte
+	}{
+		Flags:      SignFlags(0),
+		Label:      seqLabel,
+		ToBeSigned: digest,
+	}
```

Comment:
do we need an anonymous struct here? seems like local variables can just be passed

Reply:
removed anonymous structs